iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergStorageHandler.java

```diff
@@ -2095,6 +2097,55 @@ public List<Partition> getPartitionsByExpr(org.apache.hadoop.hive.ql.metadata.Ta
     }
   }
 
+  @Override
+  public List<Partition> getPartitionsWithFilter(org.apache.hadoop.hive.ql.metadata.Table hmsTable,
+      ExprNodeDesc filter, boolean currentSpec) {
+    SearchArgument sarg = ConvertAstToSearchArg.create(conf, (ExprNodeGenericFuncDesc) filter);
+    Expression exp = HiveIcebergFilterFactory.generateFilterExpression(sarg);
+    Table table = IcebergTableUtil.getTable(conf, hmsTable.getTTable());
+    int tableSpecId = table.spec().specId();
+    List<Partition> partitions = Lists.newArrayList();
+
+    TableScan scan = table.newScan().filter(exp).caseSensitive(false).includeColumnStats().ignoreResiduals();
+
+    try (CloseableIterable<FileScanTask> tasks = scan.planFiles()) {
+      tasks.forEach(task -> {
+        DataFile file = task.file();
+        PartitionSpec spec = task.spec();
+        if ((currentSpec && file.specId() == tableSpecId) || (!currentSpec && file.specId() != tableSpecId)) {
+          PartitionData partitionData = IcebergTableUtil.toPartitionData(task.partition(), spec.partitionType());
+          String partName = spec.partitionToPath(partitionData);
+          Map<String, String> partSpecMap = Maps.newLinkedHashMap();
+          Warehouse.makeSpecFromName(partSpecMap, new Path(partName), null);
+          DummyPartition partition = new DummyPartition(hmsTable, partName, partSpecMap);
+          if (!partitions.contains(partition)) {
+            partitions.add(partition);
+          }
+        }
+      });
+    } catch (IOException ioe) {
+      LOG.warn("Failed to close task iterable", ioe);
+    }
+    partitions.sort(Comparator.comparing(Partition::getName));
+    return partitions;
+  }
+
+  @Override
+  public boolean isFilterMatching(org.apache.hadoop.hive.ql.metadata.Table hmsTable, ExprNodeDesc filter) {
```

Comment:
it doesn't seem to belong here

Reply:
The goal of this method is to return true if the table contain any records matching the filter expression, false otherwise. Why it doesn't belong here?
I renamed the method to `hasDataMatchingFilterExpr`, hopefully it makes it more readable.

Reply:
Can this be moved into some util class instead of StorageHandler interface?

Reply:
This is called from hive-exec's class `AlterTableCompactOperation`.
Where can this be moved instead of StorageHandler interface to be able to access it from hive-exec?